File: frontend/lib/api/gemini.ts
// frontend/lib/api/gemini.ts
//
// Typed API client for the Gemini chat endpoint.
// Mirrors the backend schemas in app/schemas/gemini.py.
//
// Usage:
//   import { geminiApi } from "@/lib/api/gemini";
//
//   const response = await geminiApi.chat({
//     message: "Explain the concept of integration by parts.",
//     system_prompt: "You are a helpful academic study assistant.",
//   });
//   console.log(response.reply);

import { apiClient } from "./client";

// ─── Request / Response types ─────────────────────────────────────────────────

export interface ChatMessage {
  /** Must be "user" or "model" */
  role: "user" | "model";
  content: string;
}

export interface GeminiChatRequest {
  /** The current user message (required, max 8000 chars) */
  message: string;
  /** Optional system-level instruction prepended to the conversation */
  system_prompt?: string;
  /** Optional prior conversation turns for multi-turn context (max 20) */
  history?: ChatMessage[];
}

export interface GeminiChatResponse {
  /** Gemini's text reply */
  reply: string;
  /** The Gemini model variant that was used */
  model: string;
  /** Why generation stopped — e.g. "STOP" or "MAX_TOKENS" */
  finish_reason: string | null;
}

// ─── API client ───────────────────────────────────────────────────────────────

export const geminiApi = {
  /**
   * Send a message to Gemini and receive a structured reply.
   *
   * Requires authentication (Bearer token is added automatically by apiClient).
   * Access is restricted by the backend to:
   *   - Lecturers / Admins: always permitted.
   *   - Students: permitted outside of active exam sessions.
   *
   * @throws Error if the network fails, the user is unauthenticated,
   *         or Gemini returns a non-2xx status.
   */
  chat: async (request: GeminiChatRequest): Promise<GeminiChatResponse> => {
    return apiClient("/gemini/chat", {
      method: "POST",
      body: JSON.stringify(request),
    });
  },
};

File: backend/app/schemas/gemini.py
"""
app/schemas/gemini.py

Pydantic schemas for the Gemini Chat domain.
"""

from pydantic import BaseModel, Field


# ─── Chat Request ─────────────────────────────────────────────────────────────


class ChatMessage(BaseModel):
    """A single message in a multi-turn conversation history."""

    role: str = Field(..., description="Either 'user' or 'model'")
    content: str = Field(..., min_length=1, description="Message text")

    model_config = {"str_strip_whitespace": True}


class GeminiChatRequest(BaseModel):
    """
    Request body for the Gemini chat endpoint.

    Fields:
        message        — The current user message (required).
        system_prompt  — Optional system-level instructions prepended to the conversation.
        history        — Optional prior turns for multi-turn context.
    """

    message: str = Field(
        ...,
        min_length=1,
        max_length=8_000,
        description="The user's current message.",
    )
    system_prompt: str | None = Field(
        default=None,
        max_length=4_000,
        description="Optional system-level instruction for the model.",
    )
    history: list[ChatMessage] = Field(
        default_factory=list,
        max_length=20,
        description="Prior conversation turns for multi-turn context (max 20 turns).",
    )

    model_config = {"str_strip_whitespace": True}


# ─── Chat Response ────────────────────────────────────────────────────────────


class GeminiChatResponse(BaseModel):
    """
    Structured response from the Gemini chat endpoint.

    Fields:
        reply         — The model's text reply.
        model         — The Gemini model that produced this response.
        finish_reason — Why generation stopped (e.g. 'STOP', 'MAX_TOKENS').
    """

    reply: str
    model: str
    finish_reason: str | None = None
